doc/neps/nep-0034.rst

```diff
@@ -0,0 +1,131 @@
+
+When users create arrays with sequences-of-sequences, they sometimes err in
+matching the lengths of the nested sequences_, commonly called "ragged
+arrays".  Here we will refer to them as ragged nested sequences. Creating such
+arrays via ``np.array([<ragged_nested_sequence>])`` with no ``dtype`` keyword
+argument will today default to an ``object``-dtype array. Change the behaviour to
+raise a ``ValueError`` instead.
+
+Motivation and Scope
+--------------------
+
+Users who specify lists-of-lists when creating a `numpy.ndarray` via
+``np.array`` may mistakenly pass in lists of different lengths. Currently we
+accept this input and create an array with ``dtype=object``. This can be
+confusing, since it is rarely what is desired. Changing the automatic dtype
+detection to never return ``object`` for ragged nested sequences (defined as a
+recursive sequence of sequences, where not all the sequences on the same
+level have the same length) will force users who actually wish to create
+``object`` arrays to specify that explicitly. Note that lists, tuples, and
+nd.arrays are all sequences. See for instance `issue 5303`_.
+
+Usage and Impact
+----------------
+
+After this change, array creation with ragged nested sequences must explicitly
+define a dtype:
+
+    >>> np.array([[1, 2], [1]])
+    ValueError: cannot guess the desired dtype from the input
+
+    >>> np.array([[1, 2], [1]], dtype=object)
+    # succeeds, with no change from current behaviour
+
+Detailed description
+--------------------
+
+To explicitly set the shape of the object array, since it is sometimes hard to
+determine what shape is desired, one could use:
+
+    >>> arr = np.empty(correct_shape, dtype=object)
+    >>> arr[...] = values
+
+We will also reject mixed sequences of non-sequence and sequence, for instance
+all of these will be rejected:
+
+    >>> arr = np.array([np.arange(10), [10]])
+    >>> arr = np.array([[range(3), range(3), range(3)], [range(3), 0, 0]])
+
+Related Work
+------------
+
+`PR 14341`_ tried to raise an error when ragged nested sequences were specified
+with a numeric dtype ``np.array, [[1], [2, 3]], dtype=int)`` but failed due to
+false-positives, for instance ``np.array([1, np.array([5])], dtype=int)``.
+
+.. _`PR 14341`: https://github.com/numpy/numpy/pull/14341
+
+Implementation
+--------------
+
+The code to be changed is inside ``PyArray_GetArrayParamsFromObject`` and the
+internal ``discover_dimentions`` function. See `PR 14794`_.
+
+Backward compatibility
+----------------------
+
+Anyone depending on ragged nested sequences creating object arrays will need to
+modify their code. There will be a deprecation period during which the current
+behaviour will emit a ``DeprecationWarning``. There is an open question whether
+the ``assert_equal`` family of functions should be changed or users be forced
+to change code like
+
+```
+np.assert_equal(a, [[1, 2], 3])
+```
+
+to 
+
+```
+np.assert_equal(a, np.array([[1, 2], 3], dtype=object)
+```
```

Comment:
Is this decision any different to deciding whether the following should be allowed?

```python
>>> np.add([1, (2, 3)], [4, (5, 6)])
array([5, (2, 3, 5, 6)], dtype=object)
```

At any rate, it would be worth drawing attention to the fact that since this affects `asarray`, it affects almost every numpy function with list &rarr; array semantics that uses `asarray` internally.

Reply:
I think you have a point. Special-casing `array_equal` and friends would make it hard to remove the deprecation in the future. We would need to `try: ... except` around internal `asarray` use. I will drop this discussion point, users should modify their code *now* to avoid the warning.

Reply:
reformulated and moved to `Usage and Impact` for emphasis.